File: solutions/ruby/blackjack/1/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
    when 'ace'
      11
    when 'two'
      2
    when 'three'
      3
    when 'four'
      4
    when 'five'
      5
    when 'six'
      6
    when 'seven'
      7
    when 'eight'
      8
    when 'nine'
      9
    when 'ten', 'jack', 'queen', 'king'
      10
    else
      0
    end
  end

  def self.card_range(card1, card2)
    score = parse_card(card1) + parse_card(card2)

    case score
    when 4..11
      'low'
    when 12..16
      'mid'
    when 17..20
      'high'
    when 21
      'blackjack'
    end
  end

  def self.first_turn(card1, card2, dealer_card)
    score = parse_card(card1) + parse_card(card2)
    dealer_score = parse_card(dealer_card)

    if card1 == 'ace' && card2 == 'ace'
      return 'P'
    elsif score == 21 && ![10, 11].include?(dealer_score)
      'W'
    elsif score == 21 && [10, 11].include?(dealer_score)
      'S'
    elsif score <= 11 || ((12..16).include?(score) && dealer_score >= 7)
      'H'
    elsif dealer_score == 10 || ((17..20).include? score) || ((12..16).include?(score) && dealer_score < 7)
      'S'
    end
  end
end

File: solutions/java/blackjack/1/src/main/java/Blackjack.java
public class Blackjack {

    public int parseCard(String card) {
        int value = 0;

        switch (card) {
            case "ace":
                value = 11;
                break;
            case "two":
                value = 2;
                break;
            case "three":
                value = 3;
                break;
            case "four":
                value = 4;
                break;
            case "five":
                value = 5;
                break;
            case "six":
                value = 6;
                break;
            case "seven":
                value = 7;
                break;
            case "eight":
                value = 8;
                break;
            case "nine":
                value = 9;
                break;
            case "ten":
            case "jack":
            case "queen":
            case "king":
                value = 10;
                break;
            default:
                break;
        }

        return value;
    }

    public boolean isBlackjack(String card1, String card2) {
        int cardValues = parseCard(card1) + parseCard(card2);

        if(cardValues == 21) {
            return true;
        }

        return false;
    }

    public String largeHand(boolean isBlackjack, int dealerScore) {
        if(isBlackjack && dealerScore < 10) {
            return "W";
        }

        return "";
    }

    public String smallHand(int handScore, int dealerScore) {
        throw new UnsupportedOperationException("Please implement the Blackjack.smallHand method");
    }

    // FirstTurn returns the semi-optimal decision for the first turn, given the
    // cards of the player and the dealer.
    // This function is already implemented and does not need to be edited. It pulls
    // the other functions together in a
    // complete decision tree for the first turn.
    public String firstTurn(String card1, String card2, String dealerCard) {
        int handScore = parseCard(card1) + parseCard(card2);
        int dealerScore = parseCard(dealerCard);

        if (20 < handScore) {
            return largeHand(isBlackjack(card1, card2), dealerScore);
        } else {
            return smallHand(handScore, dealerScore);
        }
    }
}
